misc/mdns/src/behaviour.rs

```diff
@@ -200,28 +232,50 @@ where
                             inner: expired.into_iter(),
                         });
 
-                        return Async::Ready(NetworkBehaviourAction::GenerateEvent(event));
+                        return Poll::Ready(NetworkBehaviourAction::GenerateEvent(event));
                     }
                 },
-                Ok(Async::NotReady) => (),
-                Err(err) => warn!("tokio timer has errored: {:?}", err),
+                Poll::Pending => (),
+                Poll::Ready(Err(err)) => warn!("timer has errored: {:?}", err),
             }
         }
 
         // Polling the mDNS service, and obtain the list of nodes discovered this round.
         let discovered = loop {
-            let event = match self.service.poll() {
-                Async::Ready(ev) => ev,
-                Async::NotReady => return Async::NotReady,
+            let service = mem::replace(&mut self.service, MaybeBusyMdnsService::Poisoned);
+
+            let packet = match service {
+                MaybeBusyMdnsService::Free(service) => {
+                    self.service = MaybeBusyMdnsService::Busy(Box::pin(service.next()));
+                    continue;
+                },
+                MaybeBusyMdnsService::Busy(mut fut) => {
+                    match fut.as_mut().poll(cx) {
+                        Poll::Ready((service, packet)) => {
+                            self.service = MaybeBusyMdnsService::Free(service);
+                            packet
+                        },
+                        Poll::Pending => {
+                            self.service = MaybeBusyMdnsService::Busy(fut);
+                            return Poll::Pending;
+                        }
+                    }
+                },
+                MaybeBusyMdnsService::Poisoned => panic!("Mdns poisoned"),
             };
 
-            match event {
+            match packet {
                 MdnsPacket::Query(query) => {
-                    let _ = query.respond(
-                        params.local_peer_id().clone(),
-                        params.listened_addresses(),
-                        Duration::from_secs(5 * 60)
-                    );
+                    // MaybeBusyMdnsService should always be Free.
+                    if let MaybeBusyMdnsService::Free(ref mut service) = self.service {
+                        let resp = build_query_response(
+                            query.query_id(),
+                            params.local_peer_id().clone(),
+                            params.listened_addresses().into_iter(),
+                            MDNS_RESPONSE_TTL,
+                        );
+                        service.enqueue_response(resp.unwrap());
+                    } else { debug_assert!(false); }
```

Comment:
We could panic in release mode as well. But given that MDns is probably only a auxiliary component, I think panicing would be a bit exaggerated. In case you have suggestions how we could enforce `MaybeBusyMdnsService` to be `Free` at compile time, I am happy to do a follow up pull request.